File: src/parking/parking/scan_republisher.py
#!/usr/bin/env python3
import rclpy
from rclpy.node import Node
from sensor_msgs.msg import LaserScan
from rclpy.qos import QoSProfile, ReliabilityPolicy, HistoryPolicy

class ScanRepublisher(Node):
    def __init__(self):
        super().__init__('scan_republisher')

        # 1. Define QoS: RELIABLE (To match your LD19 Lidar)
        qos_profile = QoSProfile(
            reliability=ReliabilityPolicy.RELIABLE,
            history=HistoryPolicy.KEEP_LAST,
            depth=10
        )

        # 2. Subscribe using RELIABLE
        self.sub = self.create_subscription(
            LaserScan, 
            '/scan', 
            self.listener_callback, 
            qos_profile
        )

        # 3. Publish to Ghost Topic (Reliable)
        self.pub = self.create_publisher(LaserScan, '/scan_nav', 10)
        
        self.get_logger().info("Scan Republisher Started. Waiting for data...")
        self.count = 0

    def listener_callback(self, msg):
        # DEBUG: Print once every 40 scans (approx 2 seconds) to prove it works
        self.count += 1
        if self.count % 40 == 0:
            self.get_logger().info(f"Relaying scan {self.count} - Frame: {msg.header.frame_id} -> base_laser_nav")

        # OVERRIDE FRAME
        msg.header.frame_id = 'base_laser_nav'
        self.pub.publish(msg)

def main(args=None):
    rclpy.init(args=args)
    node = ScanRepublisher()
    rclpy.spin(node)
    node.destroy_node()
    rclpy.shutdown()

if __name__ == '__main__':
    main()

File: src/parking/src/scan_republisher.cpp
#include "rclcpp/rclcpp.hpp"
#include "sensor_msgs/msg/laser_scan.hpp"
#include <vector>
#include <algorithm> // For std::rotate

class ScanRepublisher : public rclcpp::Node
{
public:
    ScanRepublisher() : Node("scan_republisher")
    {
        // 1. Subscribe (Best Effort to match Lidar)
        auto qos_policy = rclcpp::QoS(rclcpp::KeepLast(10)).best_effort();
        
        subscription_ = this->create_subscription<sensor_msgs::msg::LaserScan>(
            "/scan", 
            qos_policy, 
            std::bind(&ScanRepublisher::topic_callback, this, std::placeholders::_1));

        // 2. Publish (Reliable for rf2o)
        publisher_ = this->create_publisher<sensor_msgs::msg::LaserScan>("/scan_nav", 10);
        
        RCLCPP_INFO(this->get_logger(), "C++ Scan Republisher Started (High Performance)");
    }

private:
    void topic_callback(sensor_msgs::msg::LaserScan::UniquePtr msg)
    {
        // 3. Frame Override
        msg->header.frame_id = "base_laser_nav";
        
        // 4. Array Rotation logic
        // We want to shift the data by +90 degrees (Quarter of the array)
        // Python equivalent: ranges = ranges[shift:] + ranges[:shift]
        
        size_t count = msg->ranges.size();
        if (count > 0) {
            size_t shift = count / 4; // 90 degrees

            // std::rotate is extremely optimized. 
            // It moves the element at 'begin + shift' to 'begin'.
            std::rotate(msg->ranges.begin(), msg->ranges.begin() + shift, msg->ranges.end());
            
            // Do the same for intensities if they exist
            if (msg->intensities.size() == count) {
                std::rotate(msg->intensities.begin(), msg->intensities.begin() + shift, msg->intensities.end());
            }
        }

        // 5. Publish (Zero-Copy)
        publisher_->publish(std::move(msg));
    }

    rclcpp::Subscription<sensor_msgs::msg::LaserScan>::SharedPtr subscription_;
    rclcpp::Publisher<sensor_msgs::msg::LaserScan>::SharedPtr publisher_;
};

int main(int argc, char * argv[])
{
    rclcpp::init(argc, argv);
    rclcpp::spin(std::make_shared<ScanRepublisher>());
    rclcpp::shutdown();
    return 0;
}